syntax = "proto2";

package ipc;

option java_package = "jp.physiome";
option java_outer_classname = "Ipc";

enum ModelLanguage {
  ISML = 0; // PHML
  SBML = 1;
  UNKNOWN = 2;
}

enum IntegrationMethod {
  EULER = 0;
  RUNGE_KUTTA = 1;
  ARK = 2;
}

message ModelVariable {

  required string key = 1;
  repeated string value = 2;

}

message ModelVariableTable {

  repeated string column = 1;
  repeated ModelVariable variable = 2;

}

message TimeUnit {

  required string name = 1;
  required int64 d = 2; // denominator
  required int64 n = 3; // numerator
  optional int32 id = 4;

}

message ModelProbeResponse {

  enum Status {
    OK = 0;
    INVALID_REQUEST = 1;
    PARSE_ERROR = 2;
    PROCESS_ERROR = 3;
    FILESYSTEM_ERROR = 4;
    MEMORY_ERROR = 5;
  }

  required Status status = 1;
  optional bytes error_message = 2;

  optional ModelLanguage language = 3;
  optional IntegrationMethod integration_method = 4;
  optional ModelVariableTable variable_table = 5;
  repeated TimeUnit time_unit = 6;
  optional string length = 7;
  optional TimeUnit length_unit = 8;
  optional string step = 9;
  optional TimeUnit step_unit = 10;
  optional int32 seed = 11;

}

message SimulationTrack {

  // the following items have the same length
  repeated string key = 1;
  repeated string name = 2;
  repeated string scope_name = 3;
  repeated string label = 4;
  repeated int32 col = 5;
  repeated int32 row = 6;

}
